File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerGeneralReductionAnnualRegularizationV2.kt
package com.amaury.pointage.v2.engine

import kotlin.math.round

/**
 * Régularise les avances mensuelles RGDU contre le droit annuel calculé.
 *
 * Sécurité : pour le cas annuel standard couvert ici, les 12 mois doivent être présents
 * explicitement, y compris les mois à 0 €. Une absence de mois n'est jamais interprétée comme 0.
 */
object EmployerGeneralReductionAnnualRegularizationV2 {
    data class MonthlyAdvance(
        val month: Int,
        val amount: Double
    )

    data class Result(
        val annualEntitlement: Double?,
        val advancesTotal: Double?,
        /** Positif = réduction complémentaire ; négatif = reprise à régulariser. */
        val adjustment: Double?,
        val reliable: Boolean,
        val warnings: List<String>
    )

    fun resolve(
        year: Int,
        annual: EmployerGeneralReductionAnnual2026V2.Result,
        monthlyAdvances: List<MonthlyAdvance>
    ): Result {
        if (year != 2026) return blocked("Régularisation RGDU : barème non intégré pour $year.")

        val malformed = monthlyAdvances.any {
            it.month !in 1..12 || !it.amount.isFinite() || it.amount < 0.0
        }
        if (malformed) {
            return blocked("Régularisation RGDU 2026 : une avance mensuelle est invalide.")
        }

        val duplicateMonths = monthlyAdvances.groupingBy { it.month }.eachCount().filterValues { it > 1 }.keys
        if (duplicateMonths.isNotEmpty()) {
            return blocked("Régularisation RGDU 2026 : plusieurs avances existent pour le même mois.")
        }

        val presentMonths = monthlyAdvances.map { it.month }.toSet()
        val expectedMonths = (1..12).toSet()
        if (presentMonths != expectedMonths) {
            val missing = (expectedMonths - presentMonths).sorted().joinToString(", ")
            return blocked(
                "Régularisation RGDU 2026 : les 12 avances mensuelles doivent être connues explicitement ; mois manquants : $missing."
            )
        }

        val annualAmount = annual.amount?.takeIf { annual.reliable && it.isFinite() && it >= 0.0 }
            ?: return blocked(
                annual.warnings.ifEmpty {
                    listOf("Régularisation RGDU 2026 : droit annuel non calculable.")
                }
            )

        val advancesTotal = roundCents(monthlyAdvances.sumOf { it.amount })
        val adjustment = roundCents(annualAmount - advancesTotal)

        return Result(
            annualEntitlement = annualAmount,
            advancesTotal = advancesTotal,
            adjustment = adjustment,
            reliable = true,
            warnings = emptyList()
        )
    }

    private fun roundCents(value: Double): Double = round(value * 100.0) / 100.0

    private fun blocked(message: String) = blocked(listOf(message))

    private fun blocked(warnings: List<String>) = Result(
        annualEntitlement = null,
        advancesTotal = null,
        adjustment = null,
        reliable = false,
        warnings = warnings.distinct()
    )
}

File: ios/HPTravail/HPTravail/SalaryV2/EmployerGeneralReductionAnnualRegularizationV2.swift
import Foundation

/// Régularise les avances mensuelles RGDU contre le droit annuel calculé.
/// Les douze mois doivent être connus explicitement, y compris les mois à 0 €.
enum EmployerGeneralReductionAnnualRegularizationV2 {
    struct MonthlyAdvance: Equatable {
        let month: Int
        let amount: Double
    }

    struct Result: Equatable {
        let annualEntitlement: Double?
        let advancesTotal: Double?
        /// Positif = réduction complémentaire ; négatif = reprise à régulariser.
        let adjustment: Double?
        let reliable: Bool
        let warnings: [String]
    }

    static func resolve(
        year: Int,
        annual: EmployerGeneralReductionAnnual2026V2.Result,
        monthlyAdvances: [MonthlyAdvance]
    ) -> Result {
        guard year == 2026 else {
            return blocked("Régularisation RGDU : barème non intégré pour \(year).")
        }

        let malformed = monthlyAdvances.contains {
            !(1...12).contains($0.month) || !$0.amount.isFinite || $0.amount < 0
        }
        guard !malformed else {
            return blocked("Régularisation RGDU 2026 : une avance mensuelle est invalide.")
        }

        let grouped = Dictionary(grouping: monthlyAdvances, by: \.month)
        guard !grouped.values.contains(where: { $0.count > 1 }) else {
            return blocked("Régularisation RGDU 2026 : plusieurs avances existent pour le même mois.")
        }

        let presentMonths = Set(monthlyAdvances.map(\.month))
        let expectedMonths = Set(1...12)
        guard presentMonths == expectedMonths else {
            let missing = expectedMonths.subtracting(presentMonths).sorted().map(String.init).joined(separator: ", ")
            return blocked("Régularisation RGDU 2026 : les 12 avances mensuelles doivent être connues explicitement ; mois manquants : \(missing).")
        }

        guard annual.reliable,
              let annualAmount = annual.amount,
              annualAmount.isFinite,
              annualAmount >= 0 else {
            return blocked(
                annual.warnings.isEmpty
                    ? ["Régularisation RGDU 2026 : droit annuel non calculable."]
                    : annual.warnings
            )
        }

        let advancesTotal = roundCents(monthlyAdvances.reduce(0) { $0 + $1.amount })
        let adjustment = roundCents(annualAmount - advancesTotal)

        return Result(
            annualEntitlement: annualAmount,
            advancesTotal: advancesTotal,
            adjustment: adjustment,
            reliable: true,
            warnings: []
        )
    }

    private static func roundCents(_ value: Double) -> Double {
        (value * 100).rounded() / 100
    }

    private static func blocked(_ warning: String) -> Result {
        blocked([warning])
    }

    private static func blocked(_ warnings: [String]) -> Result {
        Result(
            annualEntitlement: nil,
            advancesTotal: nil,
            adjustment: nil,
            reliable: false,
            warnings: unique(warnings)
        )
    }

    private static func unique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}
